lib/pacemaker/pcmk_sched_primitive.c

```diff
@@ -1339,36 +1337,39 @@ assert_role_error(pe_resource_t *rsc, pe_node_t *node, bool optional)
     CRM_ASSERT((rsc != NULL) && false);
 }
 
-gboolean
-DeleteRsc(pe_resource_t *rsc, const pe_node_t *node, gboolean optional,
-          pe_working_set_t *data_set)
+/*!
+ * \internal
+ * \brief Schedule cleanup of a resource
+ *
+ * \param[in,out] rsc       Resource to clean up
+ * \param[in]     node      Node to clean up on
+ * \param[in]     optional  Whether clean-up should be optional
+ */
+void
+pcmk__schedule_cleanup(pe_resource_t *rsc, const pe_node_t *node, bool optional)
 {
-    if (pcmk_is_set(rsc->flags, pe_rsc_failed)) {
-        pe_rsc_trace(rsc, "Resource %s not deleted from %s: failed",
-                     rsc->id, pe__node_name(node));
-        return FALSE;
+    enum pe_ordering flag = optional? pe_order_implies_then : pe_order_optional;
```

Comment:
If the *cleanup* is required, the *ordering* is optional because it's only relevant if *both* actions are required.

Conversely if the cleanup is optional, the ordering uses "implies then" so the then action will *become* required if the first action is.

That's worth a comment :)